File: macOS/Packages/ClaudeUsageCore/Sources/ClaudeUsageCore/OllamaModels.swift
import Foundation

public enum OllamaUsageSource: String, Codable, Equatable, Sendable {
    case local
    case cloud
}

/// One completed Ollama API response captured by Token Mihariban's opt-in local proxy.
/// Prompt/response text is intentionally never persisted.
public struct OllamaUsageEvent: Codable, Equatable, Sendable {
    public let timestamp: Date
    public let model: String
    public let inputTokens: Int
    public let outputTokens: Int
    public let totalDurationNanoseconds: Int64
    public let source: OllamaUsageSource
    public let requestId: String

    public var totalTokens: Int { inputTokens + outputTokens }

    public init(
        timestamp: Date,
        model: String,
        inputTokens: Int,
        outputTokens: Int,
        totalDurationNanoseconds: Int64,
        source: OllamaUsageSource,
        requestId: String
    ) {
        self.timestamp = timestamp
        self.model = model
        self.inputTokens = inputTokens
        self.outputTokens = outputTokens
        self.totalDurationNanoseconds = totalDurationNanoseconds
        self.source = source
        self.requestId = requestId
    }
}

/// Kept model-by-model even though v0.4 presents a single aggregate Ollama ring. This
/// lets a later setting expose selected models without migrating the stored history.
public struct OllamaModelBreakdown: Codable, Equatable, Sendable, Identifiable {
    public var id: String { model }
    public let model: String
    public var localInputTokens: Int
    public var localOutputTokens: Int
    public var cloudInputTokens: Int
    public var cloudOutputTokens: Int
    public var cloudEstimatedCostUSD: Double?

    public var totalTokens: Int {
        localInputTokens + localOutputTokens + cloudInputTokens + cloudOutputTokens
    }
}

public struct OllamaSnapshot: Codable, Equatable, Sendable {
    public let generatedAt: Date
    public let todayTotalTokens: Int
    public let todayLocalTokens: Int
    public let todayCloudTokens: Int
    public let todayCloudEstimatedCostUSD: Double
    public let todayModelBreakdown: [OllamaModelBreakdown]
    public let hourlyTokensToday: [HourlyUsagePoint]
    public let dailyTokensLast7Days: [DailyUsagePoint]
    public let last7DaysTotalTokens: Int
    public let dailyTokenTarget: Double
    public let colorHex: String

    public var targetFraction: Double? {
        guard dailyTokenTarget > 0 else { return nil }
        return min(1, max(0, Double(todayTotalTokens) / dailyTokenTarget))
    }

    public var hasUnpricedCloudModelToday: Bool {
        todayModelBreakdown.contains {
            $0.cloudInputTokens + $0.cloudOutputTokens > 0 && $0.cloudEstimatedCostUSD == nil
        }
    }

    public static let empty = OllamaSnapshot(
        generatedAt: Date(timeIntervalSince1970: 0),
        todayTotalTokens: 0,
        todayLocalTokens: 0,
        todayCloudTokens: 0,
        todayCloudEstimatedCostUSD: 0,
        todayModelBreakdown: [],
        hourlyTokensToday: [],
        dailyTokensLast7Days: [],
        last7DaysTotalTokens: 0,
        dailyTokenTarget: 0,
        colorHex: "#F97316"
    )
}

public enum OllamaUsageComputer {
    public static func estimatedCloudCostUSD(model: String, inputTokens: Int, outputTokens: Int) -> Double? {
        guard let pricing = OllamaCloudPricing.lookup(model: model) else { return nil }
        return Double(inputTokens) / 1_000_000 * pricing.input
            + Double(outputTokens) / 1_000_000 * pricing.output
    }

    public static func compute(
        events: [OllamaUsageEvent],
        dailyTokenTarget: Double,
        colorHex: String,
        now: Date = Date(),
        calendar: Calendar = .current
    ) -> OllamaSnapshot {
        let startOfToday = calendar.startOfDay(for: now)
        let sevenDaysAgo = calendar.date(byAdding: .day, value: -7, to: now) ?? now
        let today = events.filter { $0.timestamp >= startOfToday && $0.timestamp <= now }
        let recent = events.filter { $0.timestamp >= sevenDaysAgo && $0.timestamp <= now }

        var byModel: [String: OllamaModelBreakdown] = [:]
        for event in today {
            var row = byModel[event.model] ?? OllamaModelBreakdown(
                model: event.model,
                localInputTokens: 0,
                localOutputTokens: 0,
                cloudInputTokens: 0,
                cloudOutputTokens: 0,
                cloudEstimatedCostUSD: nil
            )
            if event.source == .cloud {
                row.cloudInputTokens += event.inputTokens
                row.cloudOutputTokens += event.outputTokens
            } else {
                row.localInputTokens += event.inputTokens
                row.localOutputTokens += event.outputTokens
            }
            byModel[event.model] = row
        }

        var cost = 0.0
        let breakdown = byModel.values.map { value -> OllamaModelBreakdown in
            var row = value
            if row.cloudInputTokens + row.cloudOutputTokens > 0,
               let estimated = estimatedCloudCostUSD(
                    model: row.model,
                    inputTokens: row.cloudInputTokens,
                    outputTokens: row.cloudOutputTokens
               ) {
                row.cloudEstimatedCostUSD = estimated
                cost += estimated
            }
            return row
        }.sorted { $0.model.localizedCaseInsensitiveCompare($1.model) == .orderedAscending }

        func hourly(_ source: [OllamaUsageEvent]) -> [HourlyUsagePoint] {
            let grouped = Dictionary(grouping: source) {
                calendar.dateInterval(of: .hour, for: $0.timestamp)?.start ?? $0.timestamp
            }
            return grouped.map { HourlyUsagePoint(hourStart: $0.key, tokens: $0.value.reduce(0) { $0 + $1.totalTokens }) }
                .sorted { $0.hourStart < $1.hourStart }
        }

        func daily(_ source: [OllamaUsageEvent]) -> [DailyUsagePoint] {
            let grouped = Dictionary(grouping: source) { calendar.startOfDay(for: $0.timestamp) }
            return grouped.map { DailyUsagePoint(day: $0.key, tokens: $0.value.reduce(0) { $0 + $1.totalTokens }) }
                .sorted { $0.day < $1.day }
        }

        let localTokens = today.filter { $0.source == .local }.reduce(0) { $0 + $1.totalTokens }
        let cloudTokens = today.filter { $0.source == .cloud }.reduce(0) { $0 + $1.totalTokens }
        return OllamaSnapshot(
            generatedAt: now,
            todayTotalTokens: localTokens + cloudTokens,
            todayLocalTokens: localTokens,
            todayCloudTokens: cloudTokens,
            todayCloudEstimatedCostUSD: cost,
            todayModelBreakdown: breakdown,
            hourlyTokensToday: hourly(today),
            dailyTokensLast7Days: daily(recent),
            last7DaysTotalTokens: recent.reduce(0) { $0 + $1.totalTokens },
            dailyTokenTarget: dailyTokenTarget,
            colorHex: colorHex
        )
    }
}

public enum OllamaUsageResponseParser {
    /// Reads the final object in either a JSON response or an NDJSON/SSE stream.
    public static func parse(
        _ data: Data,
        requestedModel: String?,
        forceCloud: Bool,
        requestId: String = UUID().uuidString,
        fallbackDate: Date = Date()
    ) -> OllamaUsageEvent? {
        var candidates: [Any] = []
        if let object = try? JSONSerialization.jsonObject(with: data) {
            candidates.append(object)
        } else if let text = String(data: data, encoding: .utf8) {
            for rawLine in text.split(whereSeparator: \.isNewline).reversed() {
                var line = String(rawLine).trimmingCharacters(in: .whitespaces)
                if line.hasPrefix("data:") { line = String(line.dropFirst(5)).trimmingCharacters(in: .whitespaces) }
                guard line != "[DONE]", let lineData = line.data(using: .utf8),
                      let object = try? JSONSerialization.jsonObject(with: lineData) else { continue }
                candidates.append(object)
            }
        }

        guard let dictionary = candidates.compactMap({ $0 as? [String: Any] }).first(where: {
            int64($0["prompt_eval_count"]) != nil || int64($0["eval_count"]) != nil
        }) else { return nil }

        let input = Int(int64(dictionary["prompt_eval_count"]) ?? 0)
        let output = Int(int64(dictionary["eval_count"]) ?? 0)
        guard input > 0 || output > 0 else { return nil }
        let model = (dictionary["model"] as? String) ?? requestedModel ?? "unknown"
        let remoteHost = dictionary["remote_host"] as? String
        let sourceText = dictionary["source"] as? String
        let isCloud = forceCloud || model.lowercased().hasSuffix(":cloud")
            || !(remoteHost ?? "").isEmpty || sourceText?.lowercased() == "cloud"
        let date = (dictionary["created_at"] as? String).flatMap(parseDate) ?? fallbackDate
        return OllamaUsageEvent(
            timestamp: date,
            model: model,
            inputTokens: input,
            outputTokens: output,
            totalDurationNanoseconds: int64(dictionary["total_duration"]) ?? 0,
            source: isCloud ? .cloud : .local,
            requestId: requestId
        )
    }

    private static func int64(_ value: Any?) -> Int64? {
        if let number = value as? NSNumber { return number.int64Value }
        if let string = value as? String { return Int64(string) }
        return nil
    }

    private static func parseDate(_ text: String) -> Date? {
        let fractional = ISO8601DateFormatter()
        fractional.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return fractional.date(from: text) ?? ISO8601DateFormatter().date(from: text)
    }
}

private enum OllamaCloudPricing {
    struct Price { let input: Double; let output: Double }

    // USD per million tokens. Kept deliberately small and easy to update as Ollama's
    // public pricing changes; unknown models remain explicitly "unpriced".
    static let models: [String: Price] = [
        "deepseek-v4-flash": Price(input: 0.44, output: 1.32),
        "deepseek-v4-pro": Price(input: 1.32, output: 3.96),
        "gemma4": Price(input: 0.14, output: 0.40),
        "glm-5.3": Price(input: 1.40, output: 4.40),
        "glm-5.3-flash": Price(input: 0.15, output: 0.50),
        "glm-5.2": Price(input: 1.40, output: 4.40),
        "glm-5.1": Price(input: 1.00, output: 3.20),
        "gpt-oss:120b": Price(input: 0.15, output: 0.60),
        "gpt-oss:20b": Price(input: 0.07, output: 0.30),
        "kimi-k3": Price(input: 3.00, output: 15.00),
        "kimi-k2.7-code": Price(input: 0.95, output: 4.00),
        "kimi-k2.6": Price(input: 0.95, output: 4.00),
        "minimax-m3": Price(input: 0.60, output: 2.40),
        "minimax-m2.7": Price(input: 0.30, output: 1.20),
        "mistral-large-3": Price(input: 0.50, output: 1.50),
        "nemotron-3-nano": Price(input: 0.06, output: 0.24),
        "nemotron-3-super": Price(input: 0.015, output: 0.60),
        "nemotron-3-ultra": Price(input: 0.10, output: 3.00),
        "qwen3.5:397b": Price(input: 0.60, output: 3.60)
    ]

    static func lookup(model: String) -> Price? {
        var key = model.lowercased()
        if key.hasSuffix(":cloud") { key.removeLast(6) }
        if key.hasSuffix("-cloud") { key.removeLast(6) }
        return models[key]
    }
}

File: TokenMihariban/Models/OllamaModels.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Text;
using System.Text.Json;

namespace TokenMihariban.Models;

public enum OllamaUsageSource { Local, Cloud }

public sealed record OllamaUsageEvent(
    DateTime Timestamp,
    string Model,
    long InputTokens,
    long OutputTokens,
    long TotalDurationNanoseconds,
    OllamaUsageSource Source,
    string RequestId)
{
    public long TotalTokens => InputTokens + OutputTokens;
}

public sealed class OllamaModelBreakdown
{
    public required string Model { get; init; }
    public long LocalInputTokens { get; set; }
    public long LocalOutputTokens { get; set; }
    public long CloudInputTokens { get; set; }
    public long CloudOutputTokens { get; set; }
    public double? CloudEstimatedCostUSD { get; set; }
    public long TotalTokens => LocalInputTokens + LocalOutputTokens + CloudInputTokens + CloudOutputTokens;
}

public sealed class OllamaSnapshot
{
    public DateTime GeneratedAt { get; init; }
    public long TodayTotalTokens { get; init; }
    public long TodayLocalTokens { get; init; }
    public long TodayCloudTokens { get; init; }
    public double TodayCloudEstimatedCostUSD { get; init; }
    public IReadOnlyList<OllamaModelBreakdown> TodayModelBreakdown { get; init; } = Array.Empty<OllamaModelBreakdown>();
    public IReadOnlyList<HourlyUsagePoint> HourlyTokensToday { get; init; } = Array.Empty<HourlyUsagePoint>();
    public IReadOnlyList<DailyUsagePoint> DailyTokensLast7Days { get; init; } = Array.Empty<DailyUsagePoint>();
    public long Last7DaysTotalTokens { get; init; }
    public double DailyTokenTarget { get; init; }
    public string ColorHex { get; init; } = "#F97316";
    public double? TargetFraction => DailyTokenTarget > 0 ? Math.Clamp(TodayTotalTokens / DailyTokenTarget, 0, 1) : null;
    public bool HasUnpricedCloudModelToday => TodayModelBreakdown.Any(x => x.CloudInputTokens + x.CloudOutputTokens > 0 && x.CloudEstimatedCostUSD is null);

    public static OllamaSnapshot Empty { get; } = new() { GeneratedAt = DateTime.UnixEpoch };
}

public static class OllamaUsageComputer
{
    public static double? EstimatedCloudCostUSD(string model, long inputTokens, long outputTokens)
    {
        return OllamaCloudPricing.TryGet(model, out var pricing)
            ? inputTokens / 1_000_000d * pricing.Input + outputTokens / 1_000_000d * pricing.Output
            : null;
    }

    public static OllamaSnapshot Compute(IReadOnlyList<OllamaUsageEvent> events, double dailyTokenTarget, string colorHex, DateTime? nowValue = null)
    {
        var now = nowValue ?? DateTime.Now;
        var todayStart = now.Date;
        var sevenDaysAgo = now.AddDays(-7);
        var today = events.Where(x => x.Timestamp.ToLocalTime() >= todayStart && x.Timestamp.ToLocalTime() <= now).ToList();
        var recent = events.Where(x => x.Timestamp.ToLocalTime() >= sevenDaysAgo && x.Timestamp.ToLocalTime() <= now).ToList();
        var breakdown = new Dictionary<string, OllamaModelBreakdown>(StringComparer.OrdinalIgnoreCase);

        foreach (var usage in today)
        {
            if (!breakdown.TryGetValue(usage.Model, out var row))
            {
                row = new OllamaModelBreakdown { Model = usage.Model };
                breakdown[usage.Model] = row;
            }
            if (usage.Source == OllamaUsageSource.Cloud)
            {
                row.CloudInputTokens += usage.InputTokens;
                row.CloudOutputTokens += usage.OutputTokens;
            }
            else
            {
                row.LocalInputTokens += usage.InputTokens;
                row.LocalOutputTokens += usage.OutputTokens;
            }
        }

        var cost = 0.0;
        foreach (var row in breakdown.Values)
        {
            if (row.CloudInputTokens + row.CloudOutputTokens == 0) continue;
            row.CloudEstimatedCostUSD = EstimatedCloudCostUSD(row.Model, row.CloudInputTokens, row.CloudOutputTokens);
            if (row.CloudEstimatedCostUSD is null) continue;
            cost += row.CloudEstimatedCostUSD.Value;
        }

        return new OllamaSnapshot
        {
            GeneratedAt = now,
            TodayLocalTokens = today.Where(x => x.Source == OllamaUsageSource.Local).Sum(x => x.TotalTokens),
            TodayCloudTokens = today.Where(x => x.Source == OllamaUsageSource.Cloud).Sum(x => x.TotalTokens),
            TodayTotalTokens = today.Sum(x => x.TotalTokens),
            TodayCloudEstimatedCostUSD = cost,
            TodayModelBreakdown = breakdown.Values.OrderBy(x => x.Model, StringComparer.OrdinalIgnoreCase).ToArray(),
            HourlyTokensToday = today.GroupBy(x => new DateTime(x.Timestamp.ToLocalTime().Year, x.Timestamp.ToLocalTime().Month, x.Timestamp.ToLocalTime().Day, x.Timestamp.ToLocalTime().Hour, 0, 0))
                .Select(x => new HourlyUsagePoint(x.Key, x.Sum(y => y.TotalTokens))).OrderBy(x => x.HourStart).ToArray(),
            DailyTokensLast7Days = recent.GroupBy(x => x.Timestamp.ToLocalTime().Date)
                .Select(x => new DailyUsagePoint(x.Key, x.Sum(y => y.TotalTokens))).OrderBy(x => x.Day).ToArray(),
            Last7DaysTotalTokens = recent.Sum(x => x.TotalTokens),
            DailyTokenTarget = dailyTokenTarget,
            ColorHex = colorHex
        };
    }
}

public static class OllamaUsageResponseParser
{
    public static OllamaUsageEvent? Parse(byte[] data, string? requestedModel, bool forceCloud, string? requestId = null, DateTime? fallbackDate = null)
    {
        var objects = new List<JsonElement>();
        try
        {
            using var doc = JsonDocument.Parse(data);
            objects.Add(doc.RootElement.Clone());
        }
        catch
        {
            var text = Encoding.UTF8.GetString(data);
            foreach (var raw in text.Split('\n', StringSplitOptions.RemoveEmptyEntries).Reverse())
            {
                var line = raw.Trim();
                if (line.StartsWith("data:", StringComparison.OrdinalIgnoreCase)) line = line[5..].Trim();
                if (line == "[DONE]") continue;
                try
                {
                    using var lineDoc = JsonDocument.Parse(line);
                    objects.Add(lineDoc.RootElement.Clone());
                }
                catch { }
            }
        }

        var root = objects.FirstOrDefault(x => x.ValueKind == JsonValueKind.Object &&
            (x.TryGetProperty("prompt_eval_count", out _) || x.TryGetProperty("eval_count", out _)));
        if (root.ValueKind != JsonValueKind.Object) return null;
        var input = ReadLong(root, "prompt_eval_count");
        var output = ReadLong(root, "eval_count");
        if (input <= 0 && output <= 0) return null;
        var model = ReadString(root, "model") ?? requestedModel ?? "unknown";
        var remoteHost = ReadString(root, "remote_host");
        var source = ReadString(root, "source");
        var isCloud = forceCloud || model.EndsWith(":cloud", StringComparison.OrdinalIgnoreCase) ||
            !string.IsNullOrWhiteSpace(remoteHost) || string.Equals(source, "cloud", StringComparison.OrdinalIgnoreCase);
        var dateText = ReadString(root, "created_at");
        var timestamp = DateTime.TryParse(dateText, CultureInfo.InvariantCulture, DateTimeStyles.RoundtripKind, out var parsed)
            ? parsed : fallbackDate ?? DateTime.UtcNow;
        return new OllamaUsageEvent(timestamp, model, input, output, ReadLong(root, "total_duration"),
            isCloud ? OllamaUsageSource.Cloud : OllamaUsageSource.Local, requestId ?? Guid.NewGuid().ToString());
    }

    private static long ReadLong(JsonElement root, string name)
    {
        if (!root.TryGetProperty(name, out var value)) return 0;
        if (value.TryGetInt64(out var number)) return number;
        return value.ValueKind == JsonValueKind.String && long.TryParse(value.GetString(), out number) ? number : 0;
    }

    private static string? ReadString(JsonElement root, string name) =>
        root.TryGetProperty(name, out var value) && value.ValueKind == JsonValueKind.String ? value.GetString() : null;
}

internal static class OllamaCloudPricing
{
    internal readonly record struct Price(double Input, double Output);

    private static readonly IReadOnlyDictionary<string, Price> Prices = new Dictionary<string, Price>(StringComparer.OrdinalIgnoreCase)
    {
        ["deepseek-v4-flash"] = new(0.44, 1.32), ["deepseek-v4-pro"] = new(1.32, 3.96),
        ["gemma4"] = new(0.14, 0.40), ["glm-5.3"] = new(1.40, 4.40),
        ["glm-5.3-flash"] = new(0.15, 0.50), ["glm-5.2"] = new(1.40, 4.40),
        ["glm-5.1"] = new(1.00, 3.20), ["gpt-oss:120b"] = new(0.15, 0.60),
        ["gpt-oss:20b"] = new(0.07, 0.30), ["kimi-k3"] = new(3.00, 15.00),
        ["kimi-k2.7-code"] = new(0.95, 4.00), ["kimi-k2.6"] = new(0.95, 4.00),
        ["minimax-m3"] = new(0.60, 2.40), ["minimax-m2.7"] = new(0.30, 1.20),
        ["mistral-large-3"] = new(0.50, 1.50), ["nemotron-3-nano"] = new(0.06, 0.24),
        ["nemotron-3-super"] = new(0.015, 0.60), ["nemotron-3-ultra"] = new(0.10, 3.00),
        ["qwen3.5:397b"] = new(0.60, 3.60)
    };

    public static bool TryGet(string model, out Price price)
    {
        var key = model.ToLowerInvariant();
        if (key.EndsWith(":cloud")) key = key[..^6];
        if (key.EndsWith("-cloud")) key = key[..^6];
        return Prices.TryGetValue(key, out price);
    }
}
